src/ray/common/asio/asio_chaos.cc

```diff
@@ -0,0 +1,103 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/asio_chaos.h"
+
+#include <cstdlib>
+#include <string_view>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_split.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace asio {
+namespace testing {
+
+namespace {
+/*
+  DelayManager is a simple chaos testing framework. Before starting ray, users
+  should set up os environment to use this feature for testing purposes.
+
+  To use this, simply do
+      export RAY_TESTING_ASIO_DELAY_MS="method1=10,method2=20"
+
+   The delay is a random number between 0 and the value. If method equals '*',
+   it will apply to all methods.
+
+   Please check warnings to make sure delay is on.
+*/
+class DelayManager {
+ public:
+  DelayManager() {
+    auto delay_env = std::getenv("RAY_TESTING_ASIO_DELAY_MS");
+    // RAY_TESTING_ASIO_DELAY="Method1=100,Method2=200"
+    if (delay_env == nullptr) {
+      return;
+    }
+    RAY_LOG(WARNING) << "RAY_TESTING_ASIO_DELAY_MS is set to be " << delay_env;
+    std::vector<std::string_view> items = absl::StrSplit(delay_env, ",");
+    for (const auto &item : items) {
+      std::vector<std::string_view> delay = absl::StrSplit(item, "=");
+      size_t delay_ms = 0;
+      if (delay.size() != 2 || !absl::SimpleAtoi(delay[1], &delay_ms)) {
+        RAY_LOG(ERROR) << "Error syntax: " << item << ", it has to be name=time";
+        continue;
+      }
+      RAY_LOG(WARNING) << "Inject asio delay on method = " << delay[0] << " for "
+                       << delay_ms << "ms";
+      if (delay[0] == "*") {
+        random_delay_ms_ = delay_ms;
+      } else {
+        delay_[std::string{delay[0]}] = delay_ms;
+      }
+    }
+  }
+
+  size_t GetMethodDelay(const std::string &name) {
+    auto it = delay_.find(name);
+    if (it == delay_.end()) {
+      return GenRandomDelay(random_delay_ms_);
+    }
+    auto actual_delay = GenRandomDelay(it->second);
+    if (actual_delay != 0) {
+      RAY_LOG_EVERY_N(DEBUG, 1000)
+          << "Delay method " << name << " for " << actual_delay << "ms";
+    }
+    return actual_delay;
+  }
+
+ private:
+  size_t GenRandomDelay(size_t delay_ms) const {
+    if (delay_ms == 0) {
+      return 0;
+    }
+    return std::rand() % delay_ms;
+  }
+
+  absl::flat_hash_map<std::string, size_t> delay_;
+  size_t random_delay_ms_ = 0;
+};
+
+static DelayManager _delay_manager;
+}  // namespace
+
+size_t get_delay_ms(const std::string &name) {
```

Comment:
Can you make the units nanoseconds? Otherwise, a lot of delays get truncated to zero / we have reduced randomness.

Reply:
```suggestion
int64_t get_delay_ns(const std::string &name) {
```

(time is not a size)

Reply:
could you share more detail about delays get truncated to zero?

here it gets the ms that we should delay running the method, so I think it won't happen right?

the reason I makes the smallest unit ms is that I'm thinking delay in nano second might not make sense since it'll need several CPU cycles to run and nano seconds seem too fine-grained.

Reply:
A millisecond is very long. Thousands of callbacks can get run or more in that time, so the quality of the delay can be increased a lot by going to us or ns.

How about at least using microseconds?